M2/Macaulay2/packages/PolyominoIdeals.m2

```diff
@@ -37,387 +74,875 @@ export {
+polyoRing = method(Options => {Field => QQ, TermOrder => Lex, RingChoice => 1})
+polyoRing CollectionOfCells := opts -> Q -> (
+    if opts.RingChoice == 1 then
+        polyoRingDefault(Q, Field => opts.Field, TermOrder => opts.TermOrder)
+    else
+        polyoRingConvex(Q, Field => opts.Field)
 );
 
--- Sub1 is a baby function which replaces 1 in the non-null entries of a generic vector.
-	
-Sub1:=(M)->(					
-	N:={};
-	for i from 0 to #M-1 do(
-		if M#i!=0 then N=join(N,{1})
-		else N=join(N,{0}); 
-	);
-	return N;
+--------------------------------------------------------------------------------------------------
+-- isPairOfInteger
+-- Checks if a list is a valid pair of integers.
+--------------------------------------------------------------------------------------------------
+isPairOfInteger = method()
+isPairOfInteger List := L -> (
+    if #L != 2 then (
+        error "Each endpoint must be a list of two integers."
+    );
+    if not all(L, x -> class x === ZZ) then (
+        error "Each coordinate must be an integer."
+    );
+    true
+)
+
+--------------------------------------------------------------------------------------------------
+-- innerInterval
+-- Checks if [a,b] forms an inner interval in a collection of cells.
+--------------------------------------------------------------------------------------------------
+innerInterval = method(Options => {})
+innerInterval(List, List, CollectionOfCells) := opts -> (a, b, Q) -> (
+    isPairOfInteger(a);
+    isPairOfInteger(b);	
+    if a#0 == b#0 or a#1 == b#1 then error "Expected a#0 < b#0 and a#1 < b#1";
+    all(toList(a#1..b#1-1), y -> all(toList(a#0..b#0-1), x -> member({x, y}, Q)))
+)
+
+--------------------------------------------------------------------------------------------------
+-- polyoIdeal
+-- Returns the ideal generated by inner 2-minors of a collection of cells.
+-- The option RingChoice with value 1 and by default returns the ideal in the ambient ring given 
+-- by polyoRingDefault. With a value different by 1 it returns the ideal in the ambient ring 
+-- given by polyoRingConvex
+--------------------------------------------------------------------------------------------------
+polyoIdeal = method (Options => {Field => QQ, TermOrder => Lex, RingChoice => 1})
+polyoIdeal CollectionOfCells := opts -> Q -> (
+    R := polyoRing(Q, 
+        Field => opts.Field, 
+        TermOrder => opts.TermOrder, 
+        RingChoice => opts.RingChoice
+    );
+    innerBinomials := {};
+    for i from 0 to #Q - 1 do (
+        A := Q#i;
+        for j from 0 to #Q - 1 do (
+            B := Q#j + {1,1}; 
+            if A#0 < B#0 and A#1 < B#1 then (
+                if innerInterval(A, B, Q) then (
+                    a := A#0;
+		    b := A#1;
+		    c := B#0;
+		    d := B#1;
+                    innerBinomials = append(innerBinomials, x_(a,b)_R * x_(c,d)_R - x_(a,d)_R * x_(c,b)_R);

```

Comment:
Calling `append` inside a loop like this isn't very efficient since we create a brand new list every single time we add a new element.

Another option would be to use `for ... list` with `continue`.  See https://macaulay2.com/doc/Macaulay2/share/doc/Macaulay2/Macaulay2Doc/html/_for.html.